rfcs/2020-03-06-1999-api-extensions-for-lua-transform.md

```diff
@@ -0,0 +1,204 @@
+
+*   **Nested Fields**
+
+    Currently accessing nested fields is possible using the dot notation:
+
+    ```lua
+    event["nested.field"] = 5
+    ```
+
+    However, users expect nested fields to be accessible as native Lua structures, for example like this:
+
+    ```lua
+    event["nested"]["field"] = 5
+    ```
+
+    See [#706](https://github.com/timberio/vector/issues/706) and [#1406](https://github.com/timberio/vector/issues/1406).
+
+*   **Setup Code**
+
+    Some scripts require expensive setup steps, for example, loading of modules or invoking shell commands. These steps should not be part of the main transform code.
+
+    For example, this code adding custom hostname
+
+    ```lua
+    if event["host"] == nil then
+      local f = io.popen ("/bin/hostname")
+      local hostname = f:read("*a") or ""
+      f:close()
+      hostname = string.gsub(hostname, "\n$", "")
+      event["host"] = hostname
+    end
+    ```
+
+    Should be split into two parts, the first part executed just once at the initialization:
+
+    ```lua
+    local f = io.popen ("/bin/hostname")
+    local hostname = f:read("*a") or ""
+    f:close()
+    hostname = string.gsub(hostname, "\n$", "")
+    ```
+
+    and the second part executed for each incoming event:
+
+    ```lua
+    if event["host"] == nil then
+      event["host"] = hostname
+    end
+    ```
+
+    See [#1864](https://github.com/timberio/vector/issues/1864).
+
+*   **Control Flow**
+
+    It should be possible to define channels for output events, similarly to how it is done in [`swimlanes`](https://vector.dev/docs/reference/transforms/swimlanes/) transform.
+
+    See [#1942](https://github.com/timberio/vector/issues/1942).
+
+## Prior Art
+
+The implementation of `lua` transform has the following design:
+
+* There is a `source` parameter which takes a string of code.
+* When a new event comes in, the global variable `event` is set inside the Lua context and the code from `source` is evaluated.
+* After that, Vector reads the global variable `event` as the processed event.
+* If the global variable `event` is set to `nil`, then the event is dropped.
+
+Events have type [`userdata`](https://www.lua.org/pil/28.1.html) with custom [metamethods](https://www.lua.org/pil/13.html), so they are views to Vector's events. Thus passing an event to Lua has zero cost, so only when fields are actually accessed the data is copied to Lua.
+
+The fields are accessed through string indexes using [Vector's dot notation](https://vector.dev/docs/about/data-model/log/#dot-notation).
+
+## Guide-level Proposal
+
+### Motivating example
+
+
+```toml
+[transforms.lua]
+  type = "lua"
+  inputs = []
+  version = "2" # defaults to 1
+  source = """
+    counter = counter + 1
+    -- without calling `emit` function nothing is produced by default
+  """
+  [transforms.lua.hooks]
+  init = """
```

Comment:
Hmmmm, I originally considered functions (Eg `start`, `stop`) but this works too!

Reply:
Notably, if we want a consistent interface for `lua`, `javascript`, and later `wasm` this may not be the best choice. It may be better to let folks specify some functions as hooks instead.

Reply:
Initially I was thinking about an API looking like this:

```toml
source = """
  function (input, emit)
    input:on_event(function (event)
      -- do something
      emit(event)
    end)
    input:on_start(function ()
      event = -- ...
      emit(event)
    end)
    input:on_stop(function ()
      -- ...
    end)
    input:on_interval(10, function()
      event = -- ...
      emit(event)
    end)
  end
"""
```

However, if such a script is a part of an already large TOML config file, such as the one from https://github.com/timberio/vector/issues/1721#issuecomment-593104785, declaring the hooks using the TOML syntax might make it easier to read and reason about. It limits generality, but in return it makes it easier to both get started and reason about scripts written by other people, especially if the user is not deeply familiar with Lua programming.

In someone needs for some reason to put functions as hooks, it is possible too:

```toml
hooks.start = """
require "mymodule"
function timer_handler(emit)
-- ...
end
my_processor = mymodule.create_processor()
"""
timers = [{ interval = 10, source = "timer_handler(emit)" }]
source = "my_processor:process(event, emit)"
```

Reply:
> However, if such a script is a part of an already large TOML config file, such as the one from #1721 (comment), declaring the hooks using the TOML syntax might make it easier to read and reason about.

In the context of the `vector.toml` file I tend to agree. But I could also see users wanting to include separate `.lua` files, which makes this point moot. For example:

```toml
source = """
require("transform.lua")
"""
```

That said, what if allowed the user to specify handlers instead of source code? Ex:

```toml
[transforms.lua]
  type = "lua"
  source = """
counter = 0

function init
  # ...
end

function process(event)
  # ...
end
"""

  handlers.init = "init" # default 
  handlers.process = "process" # default
  handlers.shutdown = "shutdown" # default 

  timers = [
    {handler = "flush", interval_secs = 10}
  ]
```

This would then allow a user to do something like:

```toml
[transforms.lua]
  type = "lua"
  source = "require('transform.lua')"

  handlers.init = "init" # default 
  handlers.process = "process" # default
  handlers.shutdown = "shutdown" # default 

  timers = [
    {handler = "flush", interval_secs = 10}
  ]
```

Reply:
@a-rodin Right, so in your example each of those scripts shares a global state? I was under the initial impression they were distinct.

Reply:
@binarylogic I think this approach suits both simple and advanced use cases. I’m going to update the RFC text.

@Hoverbear 
I’m not sure what is the best approach for Rust, but I think that it might be a solution to define a trait with methods `init`, `process`, `timers`, `shutdown` which would then be implemented by the users (or have default implementations for methods like init/shutdown hooks). So the execution model could still be the same, although the configuration would have to appear somewhat differently because it is not practical to write inline Rust in the config files.

On the other hand, as Lua is an interpreted language, for light transforms it would be possible to just write inline Lua functions with the current approach, and the trait-like approach would require writing unnecessary boilerplate code and still end up being [not idiomatic](http://lua-users.org/lists/lua-l/2017-04/msg00064.html).